Added shortcuts, implements #217

ilmich opened the pull request:
This is my final implementation.
I've upgraded database version according to latest released version of Kiss Launcher.

Cheers
Michele

Neamar commented:
(follow up for #277)

Neamar commented:
It's working good!

I'll add a few comments on the code, but really good job :)

Neamar commented on app/src/main/java/fr/neamar/kiss/DataHandler.java:
```diff
@@ -72,6 +77,11 @@ public DataHandler(Context context) {
         if (prefs.getBoolean("enable-alias", true)) {
             providers.add(new AliasProvider(context, appProvider));
         }
+        shortcutProvider = (new ShortcutProvider(context));
```
maybe add a comment explaining why we're creating a shortcut provide even if the settings is unchecked?

// We need to create the shortcutrprovider in every case, since we can get a new INSTALL_SHORTCUT intent at any time

Neamar replied:
also, unneeded parens

Neamar commented on app/src/main/java/fr/neamar/kiss/DataHandler.java:
```diff
@@ -199,6 +214,46 @@ public void addToHistory(Context context, String id) {
         DBHelper.insertHistory(context, currentQuery, id);
     }
 
+    public void addShortcut(Context context, ShortcutPojo pojo) {
```
i think we should keep the datahandler "data-agnostic" and minmize code inside. Especially, this function can be defined in broadcast/InstallShortcutHandler.java since it's the only place where it's used

Neamar commented on app/src/main/java/fr/neamar/kiss/DataHandler.java:
```diff
@@ -199,6 +214,46 @@ public void addToHistory(Context context, String id) {
+        DBHelper.insertShortcut(context, record);
+    }
+
+    public ArrayList<ShortcutPojo> getShortcuts(Context context) {
```
same thing: this can be done in doInBackground on LoadShortcutPojo (same method that's used on other Loaders)

Neamar commented on app/src/main/java/fr/neamar/kiss/DataHandler.java:
```diff
@@ -199,6 +214,46 @@ public void addToHistory(Context context, String id) {
+        return pojos;
+    }
+    
+    public void removeShortcut(Context context, String name) {
```
idem, probably best in UninstallShorcutHandler. Since it's a one-liner, can be repeated on the boradcast and on the menu item, rather than in the DataHandler class

Neamar commented on app/src/main/java/fr/neamar/kiss/broadcast/InstallShortcutHandler.java:
```diff
@@ -0,0 +1,71 @@
+        DataHandler dh = KissApplication.getDataHandler(context);
+        ShortcutProvider sp = dh.getShortcutProvider();
+
+        if (sp == null)
```
oh. So turns out you're handling the null case.

Therefore, maybe we should initialize the shortcuthandler only if `prefs.getBoolean("enable-shortcuts", true)`.

Since we're storing image in DB, i think that some app could add shorcuts too fast and we may want to givve the option to the user to "drop" all queries to add shortcuts.

I seem to remember we spoke on this earlier in the opposite sense, saying that we want the shortcuts to be added even if they are not displayed. I'm not so sure on that anymore, what's your opinion?

Neamar commented on app/src/main/java/fr/neamar/kiss/broadcast/InstallShortcutHandler.java:
```diff
@@ -0,0 +1,71 @@
+        }
+        
+        dh.addShortcut(context, pojo);
+        dh.getShortcutProvider().addShortcut(pojo);
```
sp.addShortcut(pojo)

Neamar commented on app/src/main/java/fr/neamar/kiss/broadcast/UninstallShortcutHandler.java:
```diff
@@ -0,0 +1,34 @@
+    public void onReceive(Context context, Intent data) {
+
+        DataHandler dh = KissApplication.getDataHandler(context);
+        ShortcutProvider sp = dh.getShortcutProvider();
```
if we go for the "shorcut handler can be null", don't forget to check for null here too ;)

Neamar commented on app/src/main/java/fr/neamar/kiss/broadcast/UninstallShortcutHandler.java:
```diff
@@ -0,0 +1,34 @@
+            return;
+        }
+        
+        dh.getShortcutProvider().removeShortcut(pojo);
```
sp.remove...

Neamar commented on app/src/main/java/fr/neamar/kiss/result/ShortcutResult.java:
```diff
@@ -0,0 +1,111 @@
+        Resources resources;
+        try {
+            
+            if (shortcutPojo.icon != null) {
```
you can try this from outside the try catch, before getting an instance to the packagemanager

Neamar commented:
Hi @ilmich! Any news on this one? Can't wait to publish :)

ilmich commented:
Hi @Neamar, sorry, but until tonight I did not have time to work on KISS.
But now I'm back :)
Before starting with code, if I undertand correctly you suggests to move database code from DataHandler into ShortcutProvider?!?! If yes I think also that could be the right way.

Neamar commented:
Exactly :)
Loading the POJO can be done from the Loader -- it seems easier to find and make the DataHandler smaller, which is good since it's always kept in memory.

ilmich commented:
Perfect. I've started to refactor in this way. I'll commit soon :)

Neamar commented:
I've been using the ap with shorctus enabled for the last week and noticed two weird behaviors:

* Shortcuts with icons disappear after reboot. Tested at least on Chrome: once added to home screen, the favicon appears, but on reboot the icon disappears. Very weird. Do you have the same behavior?
* At least one app is displaying twice -- once as an app, another as a shortcut, although I didn't use it as a shortcut. Looks like the shortcut was created on install? This happened on one of my other apps, https://github.com/Neamar/CineTime which makes it even weirder :p In this app, I have an [activity to create shortcuts](https://github.com/Neamar/CineTime/blob/master/app/src/main/java/fr/neamar/cinetime/ShortcutActivity.java) but haven't used it. Weird too. Did you notice this behavior with any other app?

Neamar commented:
(edit: regarding "Cinetime", i can confirm. After uninstalling it and installing again, a new shortcut appears automatically, opening the main activity ?!)

Regarding the icon issue, here is the logs:

```
10-20 15:28:19.158 1389-1917/fr.neamar.kiss D/skia: --- SkImageDecoder::Factory returned null
10-20 15:28:19.158 1389-1917/fr.neamar.kiss D/skia: --- SkImageDecoder::Factory returned null
```

ilmich commented:
@Neamar I'm using this branch for more of a week and I've notice nothing.
Regarding icons maybe there is some issue in database upgrade/creation(no icon blob field!??!) Have you tested with a clean install of kiss?!?! 
Regarding 'Cinetime' issue.. help me to find a solution. Because if installation of a new app create also a shortcut I can only try to scand AppPojos to avoid duplicates.
Same issue can apply also to contact (for example) when you create a shortcut from Contact app.

ilmich commented:
Regarding icon issue I've found this

http://stackoverflow.com/questions/27640830/bitmapfactory-returning-null

Neamar commented:
(just found another app with the same shortcut issue: "Bankin'" is creating an empty shortcut upon installation)

Neamar commented:
(regarding icons: it's a clean KISS install, tried this, same problem)

ilmich commented:
Hi @Neamar , I've fixed icon saving (it's a strange issue that I've fixed simply recompressing icon into png), refactored code and check if there is another item with same name into apps and contacts (to avoid duplicates)

Try it :)

Neamar commented:
Sorry for the delay! I'll install this right now

Neamar commented:
Some "interesting" use case:

* What happens when an app is uninstalled and a shortcut for this app is left behind? How can we handle this? (we can do nothing and expect the user to remove the shortcut manually)
* Custom icons remains after reboot \o/
* When installing some apps, I still get a "direct shortcut" (try for instance https://play.google.com/store/apps/details?id=com.bankeen, then search for "Bank")

Apart from this, looks great :) Really enjoy using it for some webpages

ilmich commented:
Hi @Neamar , what do you think about filtering shortcut only for installed apps (maybe this can be configurable)!?

Neamar commented:
I don't understand? How can you get shortcuts on non-installed apps? You
mean, if app has been uninstalled?

We can also simply remove the shortcut if it fails to launch?

On Sat, Nov 7, 2015 at 11:09 AM ilmich <notifications@github.com> wrote:

> Hi @Neamar <https://github.com/Neamar> , what do you think about
> filtering shortcut only for installed apps (maybe this can be
> configurable)!?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/Neamar/KISS/pull/293#issuecomment-154688468>.
>

ilmich commented:
@Neamar I think that best option is to remove shortcuts when app is uninstalled (that maybe is what another launchers does).
I've also noticed another little bug (when downgrading kiss version, launcher won't start because I've not implemented database downgrade)
I'll fix this soon, because I want to start another implementation (icon theming)

ilmich commented:
@Neamar I've fixed a bug on upgrading kiss app and implemented removal of shortcuts when user uninstall app.
Regarding issue with duplicated shortcuts, I've tested with suggested app (Bankin') and work for me. Can you please share catlog!??!

Regards
Michele

Neamar commented:
So, I uninstalled Bankin' and installed it again, here is the logcat:

```
11-15 12:21:34.912 25848-25848/? I/art: Late-enabling -Xcheck:jni
11-15 12:21:34.988 25848-25848/fr.neamar.kiss W/System: ClassLoader referenced unknown path: /data/app/fr.neamar.kiss-1/lib/arm
11-15 12:21:35.168 25848-25894/fr.neamar.kiss D/OpenGLRenderer: Use EGL_SWAP_BEHAVIOR_PRESERVED: true
11-15 12:21:35.304 25848-25894/fr.neamar.kiss I/Adreno-EGL: <qeglDrvAPI_eglInitialize:379>: QUALCOMM Build: 09/02/15, 76f806e, Ibddc658e36
11-15 12:21:35.308 25848-25894/fr.neamar.kiss I/OpenGLRenderer: Initialized EGL, version 1.4
11-15 12:21:36.912 25848-25876/fr.neamar.kiss I/time: 1887 milliseconds to list apps
11-15 12:21:39.511 25848-25983/fr.neamar.kiss I/time: 2596 milliseconds to list contacts
11-15 12:21:39.518 25848-26118/fr.neamar.kiss D/openCamera: Open Camera
11-15 12:21:39.675 25848-26118/fr.neamar.kiss D/openCamera: Create dummy surface texture
11-15 12:21:39.682 25848-26118/fr.neamar.kiss D/releaseCamera: Release Camera
11-15 12:21:40.380 25848-26118/fr.neamar.kiss D/releaseCamera: Release dummy surface texture
11-15 12:21:40.436 25848-26118/fr.neamar.kiss D/onUpgrade: Updating database from version 3 to version 4
11-15 12:21:40.616 25848-26118/fr.neamar.kiss I/time: 141 milliseconds to list apps
11-15 12:21:40.925 25848-25858/fr.neamar.kiss I/art: Background partial concurrent mark sweep GC freed 172306(6MB) AllocSpace objects, 0(0B) LOS objects, 39% free, 17MB/29MB, paused 1.074ms total 124.750ms
11-15 12:21:42.619 25848-26118/fr.neamar.kiss I/time: 2001 milliseconds to list contacts
11-15 12:21:53.674 25848-25894/fr.neamar.kiss E/Surface: getSlotFromBufferLocked: unknown buffer: 0x8d06d1f0
11-15 12:21:53.677 25848-25894/fr.neamar.kiss D/OpenGLRenderer: endAllStagingAnimators on 0x9d882a80 (ListPopupWindow$DropDownListView) with handle 0xaefa2c20
11-15 12:21:59.051 25848-25854/fr.neamar.kiss W/art: Suspending all threads took: 54.320ms
11-15 12:21:59.129 25848-25858/fr.neamar.kiss I/art: Background partial concurrent mark sweep GC freed 166641(5MB) AllocSpace objects, 29(600KB) LOS objects, 39% free, 18MB/30MB, paused 642us total 138.565ms
11-15 12:22:00.623 25848-25876/fr.neamar.kiss I/time: 918 milliseconds to list apps
11-15 12:22:02.807 25848-25876/fr.neamar.kiss I/time: 2183 milliseconds to list contacts
11-15 12:22:04.479 25848-25848/fr.neamar.kiss I/log: Launching app://com.android.vending/com.android.vending.AssetBrowserActivity
11-15 12:22:04.656 25848-25894/fr.neamar.kiss E/Surface: getSlotFromBufferLocked: unknown buffer: 0x8d069000
11-15 12:22:04.656 25848-25894/fr.neamar.kiss D/OpenGLRenderer: endAllStagingAnimators on 0xb4ca2400 (ListView) with handle 0xaefb1c50
11-15 12:22:04.919 25848-25848/fr.neamar.kiss W/IInputConnectionWrapper: getExtractedText on inactive InputConnection
11-15 12:22:04.995 25848-25848/fr.neamar.kiss W/IInputConnectionWrapper: getTextBeforeCursor on inactive InputConnection
11-15 12:22:05.002 25848-25848/fr.neamar.kiss W/IInputConnectionWrapper: getSelectedText on inactive InputConnection
11-15 12:22:05.003 25848-25848/fr.neamar.kiss W/IInputConnectionWrapper: getTextAfterCursor on inactive InputConnection

** HERE **
11-15 12:23:31.744 25848-25848/fr.neamar.kiss D/onReceive: Received shortcut Bankin'
11-15 12:23:31.745 25848-25848/fr.neamar.kiss D/onReceive: Received icon package name com.bankeen
11-15 12:23:31.746 25848-25848/fr.neamar.kiss D/onReceive: Received icon resource name com.bankeen:drawable/icon
** HERE **

11-15 12:23:32.085 25848-26118/fr.neamar.kiss I/time: 453 milliseconds to list apps
11-15 12:23:33.562 25848-25854/fr.neamar.kiss W/art: Suspending all threads took: 6.147ms
11-15 12:23:34.031 25848-26118/fr.neamar.kiss I/time: 1944 milliseconds to list contacts
```

Here is the screenshot:

![2015-11-15 12 25 27](https://cloud.githubusercontent.com/assets/536844/11168729/c8845986-8b9c-11e5-8f6e-ef68a48c3469.png)

Apart from this, removing the app removed the shortcut too :)

Neamar commented:
Since everything else seems fine, let's merge this one.

Maybe we'll have an idea later on why some apps create shortcuts...

Neamar commented:
Great job @ilmich :)

eevleevs commented:
Does this still work? I do not get any shortcuts on Lineage 16.

Neamar commented:
Hello!

Shortcuts were overhauled in recent versions of Android. This PR was for the old versions of shortcuts, that was much simpler.

New shortcuts are only available for the Home app -- is KISS your home app? If it is, can you open a new issue with as much information as you can provide? Thanks!
